launcher/ui/dialogs/ProgressDialog.cpp

```diff
@@ -1,29 +1,72 @@
-/* Copyright 2013-2021 MultiMC Contributors
+/// SPDX-License-Identifier: GPL-3.0-only
+/*
+ *  PrismLaucher - Minecraft Launcher
+ *  Copyright (C) 2023 Rachel Powers <508861+Ryex@users.noreply.github.com>
  *
- * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
- * You may obtain a copy of the License at
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, version 3.
  *
- *     http://www.apache.org/licenses/LICENSE-2.0
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
  *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * This file incorporates work covered by the following copyright and
+ * permission notice:
+ *
+ *      Copyright 2013-2021 MultiMC Contributors
+ *
+ *      Licensed under the Apache License, Version 2.0 (the "License");
+ *      you may not use this file except in compliance with the License.
+ *      You may obtain a copy of the License at
+ *
+ *          http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *      Unless required by applicable law or agreed to in writing, software
+ *      distributed under the License is distributed on an "AS IS" BASIS,
+ *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *      See the License for the specific language governing permissions and
+ *      limitations under the License.
  */
 
 #include "ProgressDialog.h"
 #include "ui_ProgressDialog.h"
 
+#include <limits>
 #include <QDebug>
 #include <QKeyEvent>
 
 #include "tasks/Task.h"
 
+#include "ui/widgets/SubTaskProgressBar.h"
+
+
+template<typename T> 
+int map_int_range(T value)
+{
+    // auto type_min = std::numeric_limits<T>::min();
+    auto type_min = 0;
+    auto type_max = std::numeric_limits<T>::max();
+
+    // auto int_min = std::numeric_limits<int>::min();
+    auto int_min = 0;
+    auto int_max = std::numeric_limits<int>::max();
+
+    auto type_range = type_max - type_min;
+    auto int_range = int_max - int_min;
+
+    return static_cast<int>((value - type_min) * int_range / type_range + int_min);
```

Comment:
While this isn't a bad idea by itself, it doesn't work with the scale we are dealing with here.

`int_range / type_range` evaluates to `2147483647 / 9223372036854775807 ≈ 2.328306435E−10`

Note that in this case it actually evaluates to `0`, as you are dividing two integers.

To be honest, I think we can just silently ignore values larger than `int_max`, as we would be talking about file sizes > 2 GB, which should be fairly rare. Maybe just emit a warning in the log and continue.

Reply:
(if we just ignore this map function and pass the progress values directly below, the progress bars work!)

Reply:
alternatively you could also do `current / total * 100` and pass that as the current value, while passing `100` as the new total. Or swap out 100 with a constant that defines precision.

Note that my example here is integer division, you might want to work with floats instead

Reply:
that's exactly what the fixed function does, except it used _INT_MAX instead of 100 (more precision for large downloads)